### PR TITLE
Add array_intersect

### PR DESCRIPTION
**Pull request by Pslydhh**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Add array_intersect for obtain intersect set.

### CODE REVIEW

File: be/src/exprs/vectorized/array_functions.h [hunk 1]
```diff
@@ -40,6 +40,27 @@ class ArrayFunctions {
 
 #undef DEFINE_ARRAY_DISTINCT_FN
 
+#define DEFINE_ARRAY_INTERSECT_FN(NAME, PT)                                                     \
+    static ColumnPtr array_intersect_##NAME(FunctionContext* context, const Columns& columns) { \
+        return ArrayIntersect<PT>::process(context, columns);                                   \
+    }
+
+    DEFINE_ARRAY_INTERSECT_FN(boolean, PrimitiveType::TYPE_BOOLEAN);
```
Comment (mofeiatwork) [inline be/src/exprs/vectorized/array_functions.h:48]: `APPLY_FOR_UNSIGN_OLAP_FIELD_TYPE` could be employed to reduce code for type enumeration.

File: be/src/exprs/vectorized/array_functions.tpp [hunk 2]
```diff
@@ -105,6 +105,168 @@ private:
...
+
+    class CppTypeWithOverlapTimes {
+    public:
+        CppTypeWithOverlapTimes(const CppType& item, size_t n = 0) : value(item), overlap_times(n) {}
+
+        CppType value;
+        mutable size_t overlap_times;
+    };
+
+    template <size_t type>
```
Comment (mofeiatwork) [inline be/src/exprs/vectorized/array_functions.tpp:121]: type is a trivial name?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/array_functions.tpp:121]: what type?

Comment (Pslydhh) [reply be/src/exprs/vectorized/array_functions.tpp:121]: > type is a trivial name?

Yes, I will rename this

Comment (Pslydhh) [reply be/src/exprs/vectorized/array_functions.tpp:121]: > what type?

different type is just for different column type, It's actually label name. I will refact to indicate this.

**Comment by Pslydhh** [comment]
Should change some codes after https://github.com/StarRocks/starrocks/pull/3614 merged.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

File: be/src/exprs/vectorized/array_functions.tpp [hunk 3]
```diff
@@ -524,6 +524,169 @@ private:
...
+        bool is_nullable = false;
+        bool has_null = false;
+        int null_index = 0;
+        std::vector<ArrayColumn*> src_columns;
+        src_columns.reserve(columns.size());
+        NullColumnPtr null_result = NullColumn::create();
+        null_result->resize(chunk_size);
+
+        for (int i = 0; i < columns.size(); ++i) {
+            if (columns[i]->is_nullable()) {
```
Comment (stdpain) [inline be/src/exprs/vectorized/array_functions.tpp:596]: what will happen it columns[i] was NULL literal?

File: be/src/exprs/vectorized/array_functions.tpp [hunk 4]
```diff
@@ -524,6 +524,169 @@ private:
...
+                const auto* src_nullable_column = down_cast<const NullableColumn*>(columns[i].get());
+                src_columns.emplace_back(down_cast<ArrayColumn*>(src_nullable_column->data_column().get()));
+                null_result = FunctionHelper::union_null_column(null_result, src_nullable_column->null_column());
+            } else {
+                src_columns.emplace_back(down_cast<ArrayColumn*>(columns[i].get()));
+            }
+        }
+
+        ColumnPtr src_column = ColumnHelper::unpack_and_duplicate_const_column(chunk_size, columns[null_index]);
+        ColumnPtr dest_column = src_column->clone_empty();
```
Comment (trueeyu) [inline be/src/exprs/vectorized/array_functions.tpp:610]: if the second column is nullable, the result is nullable or not ?

**Review by trueeyu — APPROVED** [review]

**Pull request merged**